# Exercise 2: TF-IDF String manipulation

In information retrieval, tf–idf (also TF*IDF, TFIDF, TF–IDF, or Tf–idf), short for term frequency–inverse document frequency, is a numerical statistic that is intended to reflect how important a word is to a document in a collection or corpus.[1] It is often used as a weighting factor in searches of information retrieval, text mining, and user modeling. The tf–idf value increases proportionally to the number of times a word appears in the document and is offset by the number of documents in the corpus that contain the word, which helps to adjust for the fact that some words appear more frequently in general. tf–idf is one of the most popular term-weighting schemes today. A survey conducted in 2015 showed that 83% of text-based recommender systems in digital libraries use tf–idf.

In [ ]:
# Turning pdb on
#pdb off
%pdb on

Automatic pdb calling has been turned ON


A. Turn paragraph into a string.


In [ ]:
tf_idf_string = """
In information retrieval, tf–idf (also TF*IDF, TFIDF, TF–IDF, or Tf–idf), short for term frequency–inverse document frequency, is a numerical statistic that is intended to reflect how important a word is to a document in a collection or corpus.[1] It is often used as a weighting factor in searches of information retrieval, text mining, and user modeling. The tf–idf value increases proportionally to the number of times a word appears in the document and is offset by the number of documents in the corpus that contain the word, which helps to adjust for the fact that some words appear more frequently in general. tf–idf is one of the most popular term-weighting schemes today. A survey conducted in 2015 showed that 83% of text-based recommender systems in digital libraries use tf–idf.
"""

B. Prints total number of words and the number of unique words.

In [ ]:
# split tf_idf_string by delimiter " " into list
words = tf_idf_string.split(" ")
print("Total number of words is", len(words))
print("Total number of unique words is", len(set(words)))

Total number of words is 102
Total number of unique words is 67


C. Which word has the most frequency?

In [ ]:
print("The most repeated word is \"" + max(set(words), key=words.count) + "\".")

The most repeated word is "a".


D. Read the TFIDF.txt file as the string and test if you will get the same results.

In [ ]:
tfidf_txt = open("/content/TFIDF.txt", "r")
tf_idf_string = tfidf_txt.read()
tfidf_txt.close()
print(tf_idf_string)

In information retrieval, tf–idf (also TF*IDF, TFIDF, TF–IDF, or Tf–idf), short for term frequency–inverse document frequency, is a numerical statistic that is intended to reflect how important a word is to a document in a collection or corpus. It is often used as a weighting factor in searches of information retrieval, text mining, and user modeling. The tf–idf value increases proportionally to the number of times a word appears in the document and is offset by the number of documents in the corpus that contain the word, which helps to adjust for the fact that some words appear more frequently in general.


Running previous code, part B produces different results. Copy and pasting had a number of 129 words and 86 unique words. The text file had 102 words and 67 unique words. Additionally, the text file version's frequency is the letter "a", as opposed to the copy paste's "the".

E. Use NLTK to remove the stop-words. What is the most frequent word now?

In [ ]:
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords

['information', 'retrieval,', 'tf–idf', '(also', 'TF*IDF,', 'TFIDF,', 'TF–IDF,', 'Tf–idf),', 'short', 'term', 'frequency–inverse', 'document', 'frequency,', 'numerical', 'statistic', 'intended', 'reflect', 'important', 'word', 'document', 'collection', 'corpus.', 'often', 'used', 'weighting', 'factor', 'searches', 'information', 'retrieval,', 'text', 'mining,', 'user', 'modeling.', 'tf–idf', 'value', 'increases', 'proportionally', 'number', 'times', 'word', 'appears', 'document', 'offset', 'number', 'documents', 'corpus', 'contain', 'word,', 'helps', 'adjust', 'fact', 'words', 'appear', 'frequently', 'general.']


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [ ]:
filtered_words = [w for w in words if not w.lower() in stopwords.words('english')]
print("The most repeated word is now \"" + max(set(filtered_words), key=words.count) + "\".")

The most repeated word is now "document".
